File: server/internal/handler/dto/sdi/rum.go
package sdi

// RUMSampleSDI is POST /api/v1/rum, one web-vitals sample from the browser.
type RUMSampleSDI struct {
	Metric string  `json:"metric" enum:"lcp,inp,cls,ttfb" description:"Tên chỉ số web-vitals" example:"lcp"`
	Value  float64 `json:"value" description:"Giá trị: mili giây cho lcp/inp/ttfb, điểm cho cls" example:"1830"`
	Route  string  `json:"route" description:"Route pattern của trang (không phải path thật)" example:"/[orgSlug]/[workspaceSlug]/tasks"`
}

File: apps/web/platform/rum.ts
import { onCLS, onINP, onLCP, onTTFB, type Metric } from "web-vitals";
import { postWebVital, type WebVitalName } from "@uniwork/core/api/endpoints/config";
import { routePattern } from "@uniwork/core/paths";

const NAMES: Record<Metric["name"], WebVitalName | undefined> = {
  LCP: "lcp", INP: "inp", CLS: "cls", TTFB: "ttfb", FCP: undefined,
};

/**
 * Report web-vitals to POST /api/v1/rum (F-11 §2.8). One coin flip per page
 * load decides whether this session reports at all; the route is the
 * pattern, never the path. Returns a no-op stop function — web-vitals has
 * no unsubscribe, so a second call after the first is guarded by the flag.
 */
export function startWebVitals(sampleRate: number): () => void {
  if (typeof window === "undefined" || sampleRate <= 0 || Math.random() >= sampleRate) return () => {};
  const report = (m: Metric) => {
    const name = NAMES[m.name];
    if (!name) return;
    void postWebVital(name, m.value, routePattern(window.location.pathname)).catch(() => {});
  };
  onLCP(report);
  onINP(report);
  onCLS(report);
  onTTFB(report);
  return () => {};
}
